airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/GeneralStagingFunctions.java

```diff
@@ -66,11 +72,23 @@ public static void copyIntoTableFromStage(final JdbcDatabase database,
                                             final List<String> stagedFiles,
                                             final String tableName,
                                             final String schemaName,
-                                            final StagingOperations stagingOperations)
+                                            final StagingOperations stagingOperations,
+                                            final String streamNamespace,
+                                            final String streamName,
+                                            final TypeAndDedupeOperationValve typerDeduperValve,
+                                            final TyperDeduper typerDeduper)
       throws Exception {
     try {
       stagingOperations.copyIntoTableFromStage(database, stageName, stagingPath, stagedFiles,
           tableName, schemaName);
+
+      AirbyteStreamNameNamespacePair streamId = new AirbyteStreamNameNamespacePair(streamNamespace, streamName);
+      if (!typerDeduperValve.containsKey(streamId)) {
+        typerDeduperValve.addStream(streamId);
+      } else if (typerDeduperValve.readyToTypeAndDedupe(streamId)) {
+        typerDeduper.typeAndDedupe(streamId.getNamespace(), streamId.getName());
```

Comment:
is this actually doing the type and dedup function? I'm not sure this is the right place to do this here.

Reply:
I would reimplement the [`DestinationFlushFunction` interface](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DestinationFlushFunction.java) and inject this [into the AsyncConsumer constructor](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/bases-destination-jdbc/src/main/java/io/airbyte/integrations/destination/staging/StagingConsumerFactory.java#L92-L100).

in this manner, we separate uploading to staging from the actual operation that flushes to the final tables.

lmk if I mis-read anything!

Reply:
also happy to talk through the code in person!

Reply:
oh actually, I think this is right, I got confused reading code. Nvm, carry on!

Reply:
yes, that's what it's doing - where would you suggest? We just want to run this after committing raw data, with some rate limit (i.e. the typerDeduperValve)

my intent in putting it here was for T+D to be a thing that runs as part of the raw commit, but we could lift it into async/serialflush, or even trigger some function/consumer from flushworkers?

Reply:
I think this is the right spot if we are eventually going to convert all async flush to type & dedup.

One assumption the async flush has today is the use of the gz file. If this is going to differ for T & Dedup, then it might be cleaner to re-implement the DestinationFlushFunction interface.

Otherwise, what we have now seems fine!